subprojects/core/src/main/java/org/gradle/execution/plan/ValuedVfsHierarchy.java

```diff
@@ -123,13 +125,10 @@ public void visitAllValues(ValueVisitor<T> valueVisitor) {
     @CheckReturnValue
     public ValuedVfsHierarchy<T> recordValue(VfsRelativePath location, T value) {
         if (location.length() == 0) {
-            return new ValuedVfsHierarchy<>(
-                ImmutableList.<T>builderWithExpectedSize(values.size() + 1)
-                    .addAll(values)
-                    .add(value)
-                    .build(), children,
-                caseSensitivity
-            );
+            List<T> newValues = new ArrayList<>(values.size() + 1);
+            newValues.addAll(values);
+            newValues.add(value);
+            return new ValuedVfsHierarchy<>(newValues, children, caseSensitivity);
```

Comment:
Can we deduplicate the code here somewhat?

Also, I don't fully understand why this is creating less garbage. All we save here is the single short-lived `Builder` instance, no?